segregation/inference/comparative.py

```diff
@@ -0,0 +1,440 @@
+
+
+def _prepare_comparative_data(df1, df2, group_pop_var1, group_pop_var2, total_pop_var1, total_pop_var2):
+    df1 = df1.copy()
+    df2 = df2.copy()
+    if hasattr(df1, "geometry"):
+        df1 = df1[[group_pop_var1, total_pop_var1, df1.geometry.name]]
+    else:
+        df1 = df1[[group_pop_var1, total_pop_var1]]
+
+    if hasattr(df2, "geometry"):
+        df2 = df2[[group_pop_var2, total_pop_var2, df2.geometry.name]]
+    else:
+        df2 = df2[[group_pop_var2, total_pop_var2]]
+
+    return df1, df2
+
+
+
+def _generate_counterfactual(
+    data1,
+    data2,
+    group_pop_var1,
+    total_pop_var1,
+    group_pop_var2,
+    total_pop_var2,
+    counterfactual_approach="composition",
+):
+    """Generate a counterfactual variables.
+
+    Given two contexts, generate counterfactual distributions for a variable of
+    interest by simulating the variable of one context into the spatial
+    structure of the other.
+
+    Parameters
+    ----------
+    data1 : pd.DataFrame or gpd.DataFrame
+        Pandas or Geopandas dataframe holding data for context 1
+
+    data2 : pd.DataFrame or gpd.DataFrame
+        Pandas or Geopandas dataframe holding data for context 2
+
+    group_pop_var : str
+        The name of variable in both data that contains the population size of the group of interest
+
+    total_pop_var : str
+        The name of variable in both data that contains the total population of the unit
+
+    approach : str, ["composition", "share", "dual_composition"]
+        Which approach to use for generating the counterfactual.
+        Options include "composition", "share", or "dual_composition"
+
+    Returns
+    -------
+    two DataFrames
+        df1 and df2  with appended columns 'counterfactual_group_pop', 'counterfactual_total_pop', 'group_composition' and 'counterfactual_composition'
+
+    """
+    df1, df2 = DUAL_SIMULATORS[counterfactual_approach](
+        data1, data2, group_pop_var1, total_pop_var1, group_pop_var2, total_pop_var2,
+    )
+    df1["group_composition"] = (df1[group_pop_var1] / df1[total_pop_var1]).fillna(0)
+    df2["group_composition"] = (df2[group_pop_var2] / df2[total_pop_var2]).fillna(0)
+
+    df1["counterfactual_composition"] = (
+        df1["counterfactual_group_pop"] / df1["counterfactual_total_pop"]
+    ).fillna(0)
+    df2["counterfactual_composition"] = (
+        df2["counterfactual_group_pop"] / df2["counterfactual_total_pop"]
+    ).fillna(0)
+
+    df1 = df1.drop(columns=[group_pop_var1, total_pop_var1], axis=1)
+    df2 = df2.drop(columns=[group_pop_var2, total_pop_var2], axis=1)
+
+    return df1, df2
+
+
+def sim_composition(
+    df1, df2, group_pop_var1, total_pop_var1, group_pop_var2, total_pop_var2,
+):
+    """Simulate the spatial distribution of a population group in a region using the CDF of a comparison region.
+
+    For each spatial unit i in region 1, take the unit's rank in the distribution, and swap the group composition
```

Comment:
These shouldn't be based on ranks, but rather percentiles from the respective distributions.